File: framework/languages/node/e2e/RuntimeMonitoring/Client/Scenarios/mon-a5-fixed-kinds-scenario.ts
// MON-A5: Store 장애와 복구 상태를 관찰한다 시나리오를 검증한다.
import { spawn } from 'node:child_process';
import type { EvidenceWaitReq, ProfileReq, ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../../../http-client';
import { routeStatusesFromEvidence, waitForRouteStatus, delay } from '../Support/public-status';
import { ensure } from '../Support/scenario-assert';

export async function runMonA5(options: ClientOptions): Promise<void> {
  const before = await waitForRouteStatus(
    options.serviceUrl,
    (status) => status.isReady && status.channels.some((channel) => channel.isReady),
    'MON-A5 RouteMesh was not ready before the Store outage.'
  );

  await docker('pause', options.redisContainer);
  try {
    const degraded = await waitForRouteStatus(
      options.serviceUrl,
      (status) => !status.isReady && (status.state === 2 || status.placement.unavailableReason === 3),
      'MON-A5 RouteMesh did not expose Store unavailability.'
    );
    ensure(BigInt(degraded.sequence) >= BigInt(before.sequence), 'MON-A5 RouteMesh sequence moved backwards during Store outage.');
    const observed = routeStatusesFromEvidence(await getJson<readonly string[]>(options.serviceUrl, '/evidence'));
    ensure(
      observed.some((status) => !status.isReady && (status.state === 2 || status.placement.unavailableReason === 3)),
      'MON-A5 public observer did not expose the degraded Store status.'
    );
  } finally {
    await docker('unpause', options.redisContainer);
  }

  const restored = await waitForRouteStatus(
    options.serviceUrl,
    (status) => status.isReady && status.channels.some((channel) => channel.isReady),
    'MON-A5 RouteMesh did not recover after the Store was resumed.'
  );
  ensure(BigInt(restored.sequence) > BigInt(before.sequence), 'MON-A5 RouteMesh sequence did not advance after Store recovery.');
  const reply = await postJson<ProfileRes>(options.triggerUrl, '/profile/request/service-b', {
    value: 'after-recovery',
    marker: 'mon-a5-after-recovery'
  } satisfies ProfileReq);
  ensure(reply.providerRid === 'svc-b' && reply.value === 'profile:after-recovery', 'MON-A5 request did not recover after Store restoration.');
  await postJson<string[]>(options.serviceBUrl, '/evidence/wait', {
    containsAll: ['profile-request|rid=svc-b|marker=mon-a5-after-recovery|value=after-recovery'],
    containsAnyGroups: [],
    timeoutMilliseconds: 10000
  } satisfies EvidenceWaitReq);
  console.log('scenario MON-A5 passed');
}

async function docker(verb: 'pause' | 'unpause', container: string): Promise<void> {
  await new Promise<void>((resolve, reject) => {
    const child = spawn('docker', [verb, container], { stdio: 'ignore' });
    child.once('exit', (code) => {
      if (code === 0) resolve();
      else reject(new Error(`docker ${verb} ${container} failed with exit code ${code}`));
    });
    child.once('error', reject);
  });
  await delay(100);
}

File: framework/languages/dotnet/e2e/RuntimeMonitoring/Client/Scenarios/MonA5FixedKindsScenario.cs
// Verifies Redis location health events and admitted-route continuity.
using System.Diagnostics;
using RuntimeMonitoring.Client.Support;
using RuntimeMonitoring.Shared;
using Zlink.HttpClient;

namespace RuntimeMonitoring.Client.Scenarios;

internal static class MonA5FixedKindsScenario
{
    public static async Task RunAsync(ClientOptions options)
    {
        using var service = ZLinkHttpClient.Create(options.ServiceUrl)
            .Timeout(TimeSpan.FromSeconds(35))
            .Build();
        var baseline = await SnapshotAsync(service);
        var baselineEvidence =
            (await service.Get("/evidence").Async<string[]>()).Body.Length;
        ZlinkStreamAssert.Ensure(
            baseline.IsReady
            && baseline.Peers.Any(peer =>
                peer.Rid.StartsWith("svc-b-", StringComparison.Ordinal)
                && peer.State == "Ready"),
            "MON-A5 baseline RouteMesh status was not ready.");

        await RunDockerAsync("pause", options.RedisContainer);
        try
        {
            await WaitForLocationEventAsync(service, baselineEvidence, "degraded");
            var duringOutage = await SnapshotAsync(service);
            ZlinkStreamAssert.Ensure(
                duringOutage.IsReady
                && duringOutage.Channels.Single(channel =>
                    channel.ChannelName == RuntimeMonitoringNames.Channel).IsReady,
                "MON-A5 store outage incorrectly removed the admitted messaging path.");

            var request = await service.Post("/profile/request")
                .Body(new ProfileReq("during-outage", "mon-a5-during-outage"))
                .Async<ProfileRes>();
            ZlinkStreamAssert.Ensure(
                request.Body.Marker == "mon-a5-during-outage",
                "MON-A5 admitted messaging did not continue during the store outage.");
        }
        finally
        {
            await RunDockerAsync("unpause", options.RedisContainer);
        }

        await WaitForLocationEventAsync(service, baselineEvidence, "ready");
        var recovered = await SnapshotAsync(service);
        ZlinkStreamAssert.Ensure(
            recovered.IsReady
            && recovered.Peers.Any(peer =>
                peer.Rid.StartsWith("svc-b-", StringComparison.Ordinal)
                && peer.State == "Ready"),
            "MON-A5 ready topology did not remain available after store recovery.");

        Console.WriteLine("scenario MON-A5 passed");
    }

    private static async Task<MeshRuntimeSnapshotRes> SnapshotAsync(
        ZLinkHttpClient service)
        => (await service.Get(
                $"/runtime/snapshot/{RuntimeMonitoringNames.Channel}")
            .Async<MeshRuntimeSnapshotRes>()).Body;

    private static async Task WaitForLocationEventAsync(
        ZLinkHttpClient service,
        int afterIndex,
        string state)
    {
        var evidence = (await service.Post("/evidence/wait")
            .Body(new EvidenceWaitReq(
                [
                    $"source={RuntimeMonitoringNames.LocationRuntimeSource}",
                    "identifier=zlink.runtime.location.store_changed",
                    $"reason={state}"
                ],
                [],
                TimeoutMilliseconds: 3000,
                AfterIndex: afterIndex))
            .Async<string[]>()).Body;
        ZlinkStreamAssert.Ensure(
            evidence.Any(line =>
                line.Contains(
                    "identifier=zlink.runtime.location.store_changed",
                    StringComparison.Ordinal)
                && line.Contains($"reason={state}", StringComparison.Ordinal)),
            $"MON-A5 '{state}' location event was missing.");
    }

    private static async Task RunDockerAsync(string verb, string container)
    {
        var start = new ProcessStartInfo("docker")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false
        };
        start.ArgumentList.Add(verb);
        start.ArgumentList.Add(container);
        using var process = Process.Start(start)
                            ?? throw new InvalidOperationException(
                                $"Failed to run docker {verb}.");
        await process.WaitForExitAsync();
        if (process.ExitCode == 0)
            return;
        var error = await process.StandardError.ReadToEndAsync();
        throw new InvalidOperationException(
            $"docker {verb} {container} failed: {error}");
    }
}
